### PR TITLE
build(toolchain): bump to 2022-10-16

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

### CODE REVIEW

File: src/common/src/error.rs [hunk 1]
```diff
@@ -230,7 +230,9 @@ impl Debug for RwError {
             "{}\n{}",
             self.inner,
             // Use inner error's backtrace by default, otherwise use the generated one in `From`.
-            self.inner.backtrace().unwrap_or(&*self.backtrace)
+            (&self.inner as &dyn std::error::Error)
+                .request_ref::<Backtrace>()
+                .unwrap_or(&*self.backtrace)
```
Comment (BugenZhao) [inline src/common/src/error.rs:235]: 🤔

Comment (TennyZhuang) [reply src/common/src/error.rs:235]: https://github.com/rust-lang/rust/issues/99301#issuecomment-1237530400

**Comment by xxchan** [comment]
The ICE is fixed. Now we have TAIT bug 🤪 https://github.com/rust-lang/rust/issues/101750

**Comment by BugenZhao** [comment]
Really want https://github.com/rust-lang/cargo/pull/11114 after bumping. 🤤

File: src/storage/src/lib.rs [hunk 2]
```diff
@@ -42,6 +39,9 @@
 #![feature(once_cell)]
 #![cfg_attr(coverage, feature(no_coverage))]
 #![recursion_limit = "256"]
+#![feature(error_generic_member_access)]
+#![feature(provide_any)]
+#![expect(clippy::result_large_err, reason = "FIXME: HummockError is large")]
```
Comment (BugenZhao) [inline src/storage/src/lib.rs:44]: Should we open an issue for this? 🤔

Comment (xxchan) [inline src/storage/src/lib.rs:44]: notable

File: src/meta/src/lib.rs [hunk 3]
```diff
@@ -12,24 +12,28 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#![feature(backtrace)]
+#![expect(clippy::iter_kv_map, reason = "FIXME: fix later")]
+#![expect(
+    clippy::or_fun_call,
```
Comment (BugenZhao) [inline src/meta/src/lib.rs:17]: 👀

File: src/common/src/types/chrono_wrapper.rs [hunk 4]
```diff
@@ -116,7 +116,7 @@ impl NaiveTimeWrapper {
     pub fn with_secs_nano_value(secs: u32, nano: u32) -> value_encoding::Result<Self> {
         Ok(NaiveTimeWrapper::new(
             NaiveTime::from_num_seconds_from_midnight_opt(secs, nano)
-                .ok_or(ValueEncodingError::InvalidNaiveTimeEncoding(secs, nano))?,
+                .ok_or_else(|| ValueEncodingError::InvalidNaiveTimeEncoding(secs, nano))?,
```
Comment (BugenZhao) [inline src/common/src/types/chrono_wrapper.rs:119]: There shouldn't be any difference for enum construction.

**Review by BugenZhao — APPROVED** [review]

**Review by skyzh — APPROVED** [review]

File: src/storage/src/lib.rs (hunk 2 above)
Comment (xxchan) [reply src/storage/src/lib.rs:44]: opened

File: src/common/src/types/chrono_wrapper.rs (hunk 4 above)
Comment (xxchan) [reply src/common/src/types/chrono_wrapper.rs:119]: https://github.com/rust-lang/rust-clippy/issues/9663

**Comment by skyzh** [comment]
Build of `risingwave_stream` will stuck on macOS. Still investigating why.

**Comment by skyzh** [comment]
```
INFO rustc_trait_selection::traits::query::normalize normalize::<rustc_middle::mir::ConstantKind>: result=Ok(Val(ZeroSized, for<'a, 'b, 'c> fn(std::pin::Pin<&'a mut futures_async_stream::try_stream::from_generator::GenTryStream<[static generator@src/stream/src/executor/hash_join.rs:760:5: 760:61]>>, &'b mut std::task::Context<'c>) -> std::task::Poll<std::option::Option<<futures_async_stream::try_stream::from_generator::GenTryStream<[static generator@src/stream/src/executor/hash_join.rs:760:5: 760:61]> as tokio_stream::Stream>::Item>> {<futures_async_stream::try_stream::from_generator::GenTryStream<[static generator@src/stream/src/executor/hash_join.rs:760:5: 760:61]> as tokio_stream::Stream>::poll_next})) with 4052 obligations
INFO rustc_trait_selection::traits::query::normalize normalize::<rustc_middle::ty::subst::GenericArg>: result=Ok(unsafe fn(&mut futures_async_stream::try_stream::from_generator::GenTryStream<[static generator@src/stream/src/executor/hash_join.rs:760:5: 760:61]>) -> std::pin::Pin<&mut futures_async_stream::try_stream::from_generator::GenTryStream<[static generator@src/stream/src/executor/hash_join.rs:760:5: 760:61]>> {std::pin::Pin::<&mut futures_async_stream::try_stream::from_generator::GenTryStream<[static generator@src/stream/src/executor/hash_join.rs:760:5: 760:61]>>::new_unchecked}) with 4052 obligations
INFO rustc_trait_selection::traits::query::normalize normalize::<rustc_middle::mir::ConstantKind>: result=Ok(Val(ZeroSized, unsafe fn(&mut futures_async_stream::try_stream::from_generator::GenTryStream<[static generator@src/stream/src/executor/hash_join.rs:760:5: 760:61]>) -> std::pin::Pin<&mut futures_async_stream::try_stream::from_generator::GenTryStream<[static generator@src/stream/src/executor/hash_join.rs:760:5: 760:61]>> {std::pin::Pin::<&mut futures_async_stream::try_stream::from_generator::GenTryStream<[static generator@src/stream/src/executor/hash_join.rs:760:5: 760:61]>>::new_unchecked})) with 4052 obligations
```

It seems that Rust is very slow with resolving futures_async_stream. It took about 1sec for each resolving.

**Comment by skyzh** [comment]
To reproduce, `RUSTC_LOG=debug ./risedev d`

**Comment by BugenZhao** [comment]
Seems there's some new false positives for resolving lifetime failure. 🤣

**Comment by skyzh** [comment]
As a comparison, previous builds only have 0 obligations for all these things.

**Comment by skyzh** [comment]
future_async_stream requires Generator, which seems to have performance issues. Probably related to https://github.com/rust-lang/rust/pull/101692. I can test if that PR would solve this issue by manually build a Rust compiler with that branch...

**Comment by skyzh** [comment]
Also the lifetime issue doesn't look like a false positive.

```
        let rotate_last_mut = |buffers: &'a mut Vec<_>| {
            buffers.push(DioBuffer::with_capacity_in(
                self.buffer_capacity,
                &DIO_BUFFER_ALLOCATOR,
            ));
            buffers.last_mut().unwrap()
        };
```

is defined for `'a mut Vec`

where

```
rotate_last_mut(&mut self.buffers),
```

is following the lifetime of `&mut self`. `buffers`' lifetime has no relationship to `'a`. The compiler's report seems correct.

cc @MrCroxx would you please help fix this?

**Comment by skyzh** [comment]
Unluckily, https://github.com/rust-lang/rust/pull/101692 doesn't help much 🤣 Probably we will need to help the Rust community to fix this...

* futures_async_stream seems to generate too much more obligations than before.
* need fix lifetime issues
* may merge this PR without bumping the toolchain

**Comment by skyzh** [comment]
Okay, the problem comes to HashJoin again. After clearing all function body of `HashJoin`:

```
#[try_stream(ok = Message, error = StreamExecutorError)]
async fn into_stream(mut self) {}
```

Everything works!

@yuhao-su are you still maintaining the streaming hash join component? Do you have time to look into why HashJoin is causing rustc to stuck?

**Comment by yuhao-su** [comment]
> Do you have time to look into why HashJoin is causing rustc to stuck?

I'll take a look.

**Comment by skyzh** [comment]
Unluckily, the bombing obligations seem to originate from StateTableIter

```
INFO rustc_trait_selection::traits::query::normalize normalize::<rustc_middle::ty::subst::GenericArg>: result=Ok(std::future::from_generator::GenFuture<[static generator@src/stream/src/common/mod.rs:31:45: 37:2]>) with 4052 obligations
INFO rustc_trait_selection::traits::query::normalize normalize::<rustc_middle::ty::subst::GenericArg>: result=Ok(std::future::from_generator::GenFuture<[static generator@risingwave_storage::table::streaming_table::state_table::StateTable<risingwave_storage::monitor::MonitoredStateStore<risingwave_storage::hummock::HummockStorage>>::iter::{closure#0}]>) with 4052 obligations
INFO rustc_trait_selection::traits::query::normalize normalize::<rustc_middle::ty::subst::GenericArg>: result=Ok(std::future::from_generator::GenFuture<[static generator@risingwave_storage::table::streaming_table::state_table::StateTable<risingwave_storage::monitor::MonitoredStateStore<risingwave_storage::hummock::HummockStorage>>::iter_with_pk_prefix::{closure#0}]>) with 4052 obligations
```

which cases all executors to have a lot of obligations, whereas hash join needs a lot of specialization, and cases the most problems.

**Comment by skyzh** [comment]
After changing `iter_key_and_val` to `boxed`, risingwave_stream compiles in 40 seconds (still seems slow). cc @BugenZhao now it's your turn to investigate what's happening 🥵

```
    pub async fn iter_key_and_val<'a>(
        &'a self,
        pk_prefix: &'a Row,
    ) -> StorageResult<RowStreamWithPk<'a, S>> {
        let (mem_table_iter, storage_iter_stream) = self
            .iter_with_pk_prefix_inner(pk_prefix, self.epoch())
            .await?;
        let storage_iter = storage_iter_stream.into_stream().boxed();

        Ok(
            StateTableRowIter::new(mem_table_iter, storage_iter, self.row_deserializer.clone())
                .into_stream().boxed(),
        )
    }
```

**Comment by skyzh** [comment]
My wild guess is that while rustc allows more lifetime generics in the latest version, some wrongly-labeled lifetimes will cause compile issues. Every time we add `'a` at `&self`, there must be something going wrong.

**Comment by skyzh** [comment]
The issue now narrows down to risingwave_storage. It seems that it produces 4052 obligations...

image

Most of them are `OutlivesPredicate` (lifetime issues)

**Comment by TennyZhuang** [comment]
turn to @wenym1

**Comment by TennyZhuang** [comment]
IIRC storage have many unnecessary generics, but I’m not sure whether it’s related.

**Comment by skyzh** [comment]
I feel this issue more related to StateTable implementation than HummockStorage itself. The obligations will only bomb when processing StateTable-related code.

**Comment by BugenZhao** [comment]
Will implementing state table iterator manually with `poll_next` help?

**Comment by skyzh** [comment]
I'm not sure, but I'd suggest investigating where's the obligations from.

**Comment by MrCroxx** [comment]
> Also the lifetime issue doesn't look like a false positive.
> 
> ```
>         let rotate_last_mut = |buffers: &'a mut Vec<_>| {
>             buffers.push(DioBuffer::with_capacity_in(
>                 self.buffer_capacity,
>                 &DIO_BUFFER_ALLOCATOR,
>             ));
>             buffers.last_mut().unwrap()
>         };
> ```
> 
> is defined for `'a mut Vec`
> 
> where
> 
> ```
> rotate_last_mut(&mut self.buffers),
> ```
> 
> is following the lifetime of `&mut self`. `buffers`' lifetime has no relationship to `'a`. The compiler's report seems correct.
> 
> cc @MrCroxx would you please help fix this?

Sure. I'll handle it.

**Comment by lmatz** [comment]
https://github.com/risingwavelabs/risingwave/issues/1334 can be closed after this

**Comment by skyzh** [comment]
I'll take over this PR and get it merged with my pace.

**Comment by victoryang00** [comment]
Thank you guys very much

**Comment by skyzh** [comment]
close in favor of https://github.com/risingwavelabs/risingwave/pull/6025

**Pull request closed without merging**